### ISSUE TITLE
groupby_rolling negative vs non-negative offsets behavior differs

### ISSUE THREAD

**Issue by swpease** [pola-rs/polars#6622]
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

The default `offset` of `-period`, with the default `closed = "right"`, results in each window including rows, for a given index (index - period, index]. 

I tried setting `offset = 0d`, and the window includes [index, index + period]. This makes sense from the standpoint of "why would you want to exclude the current index from your aggregation on it?", but is what I would expect if `closed = "both"`.

Setting positive `offset`s results in some unusual behavior as well (see below). Again, I'm not sure *why* someone would want to do it.

#### Zero Offset
**[Example 1]** 
#### Positive Offsets
**[Example 2]** If you make an offset >= the minimum timedelta in your index, the earlier of those two indexes gets an `agg` output of `null`.
**[Example 3]** This just confuses me.

### Reproducible example

```python
# Example 1
from datetime import datetime
df = pl.DataFrame(
    {
        "time": [
            datetime(2021, 1, 10),
            datetime(2021, 1, 11),
            datetime(2021, 1, 12),
        ],
        "val": range(1, 4),
    }
)
(df
    .groupby_rolling("time",
                     period="1d",
                     offset="0d",
                    )
    .agg([
        pl.col("val").sum(),
        pl.col("time").min().alias("min_date"),
        pl.col("time").max().alias("max_date"),
    ])
)
#┌─────────────────────┬──────┬─────────────────────┬─────────────────────┐
#│ time                ┆ val  ┆ min_date            ┆ max_date            │
#│ ---                 ┆ ---  ┆ ---                 ┆ ---                 │
#│ datetime[μs]        ┆ i64  ┆ datetime[μs]        ┆ datetime[μs]        │
#╞═════════════════════╪══════╪═════════════════════╪═════════════════════╡
#│ 2021-01-10 00:00:00 ┆ 3    ┆ 2021-01-10 00:00:00 ┆ 2021-01-11 00:00:00 │
#│ 2021-01-11 00:00:00 ┆ 5    ┆ 2021-01-11 00:00:00 ┆ 2021-01-12 00:00:00 │
#│ 2021-01-12 00:00:00 ┆ null ┆ null                ┆ null                │
#└─────────────────────┴──────┴─────────────────────┴─────────────────────┘

# Example 2
import polars as pl
from datetime import datetime
df = pl.DataFrame(
    {
        "time": [
            datetime(2021, 2, 25),
            datetime(2021, 2, 28),
            datetime(2021, 3, 11),
        ],
        "val": range(1, 4),
    }
)

(df
    .groupby_rolling("time",
                     period="1mo",
                    #  offset="2d23h59m59s",  # Switch to here and it works.
                     offset="3d",
                    )
    .agg([
        pl.col("val").sum(),
    ])
)
#┌─────────────────────┬──────┐
#│ time                ┆ val  │
#│ ---                 ┆ ---  │
#│ datetime[μs]        ┆ i64  │
#╞═════════════════════╪══════╡
#│ 2021-02-25 00:00:00 ┆ null │
#│ 2021-02-28 00:00:00 ┆ 5    │
#│ 2021-03-11 00:00:00 ┆ null │
#└─────────────────────┴──────┘


# Example 3
from datetime import datetime
df = pl.DataFrame(
    {
        "time": [
            datetime(2021, 1, 10),
            datetime(2021, 1, 11),
            datetime(2021, 1, 12),
            datetime(2021, 1, 13),
            datetime(2021, 1, 14),
            datetime(2021, 1, 15),
            datetime(2021, 1, 16),
        ],
        "val": range(1, 8),
    }
)
(df
    .groupby_rolling("time",
                     period="1d",
                     offset="1d",
                    )
    .agg([
        pl.col("val").sum(),
        pl.col("time").min().alias("min_date"),
        pl.col("time").max().alias("max_date"),
    ])
)

#  time                ┆ val  ┆ min_date            ┆ max_date            │
#│ ---                 ┆ ---  ┆ ---                 ┆ ---                 │
#│ datetime[μs]        ┆ i64  ┆ datetime[μs]        ┆ datetime[μs]        │
#╞═════════════════════╪══════╪═════════════════════╪═════════════════════╡
#│ 2021-01-10 00:00:00 ┆ null ┆ null                ┆ null                │
#│ 2021-01-11 00:00:00 ┆ null ┆ null                ┆ null                │
#│ 2021-01-12 00:00:00 ┆ 12   ┆ 2021-01-12 00:00:00 ┆ 2021-01-14 00:00:00 │
#│ 2021-01-13 00:00:00 ┆ 15   ┆ 2021-01-13 00:00:00 ┆ 2021-01-15 00:00:00 │
#│ 2021-01-14 00:00:00 ┆ null ┆ null                ┆ null                │
#│ 2021-01-15 00:00:00 ┆ null ┆ null                ┆ null                │
#│ 2021-01-16 00:00:00 ┆ null ┆ null                ┆ null                │
#└─────────────────────┴──────┴─────────────────────┴─────────────────────┘
```

### Expected behavior

I would expect Example 1 to produce:
```
┌─────────────────────┬──────┬─────────────────────┬─────────────────────┐
│ time                ┆ val  ┆ min_date            ┆ max_date            │
│ ---                 ┆ ---  ┆ ---                 ┆ ---                 │
│ datetime[μs]        ┆ i64  ┆ datetime[μs]        ┆ datetime[μs]        │
╞═════════════════════╪══════╪═════════════════════╪═════════════════════╡
│ 2021-01-10 00:00:00 ┆ 2    ┆ 2021-01-10 00:00:00 ┆ 2021-01-11 00:00:00 │
│ 2021-01-11 00:00:00 ┆ 3    ┆ 2021-01-11 00:00:00 ┆ 2021-01-12 00:00:00 │
│ 2021-01-12 00:00:00 ┆ null ┆ null                ┆ null                │
└─────────────────────┴──────┴─────────────────────┴─────────────────────┘
```

I'm not sure about the positive offset examples, except that the first `null`s are unexpected.

### Installed versions

<details>

```
---Version info---
Polars: 0.16.1
Index type: UInt32
Platform: macOS-12.5-arm64-arm-64bit
Python: 3.11.1 (v3.11.1:a7a450f84a, Dec  6 2022, 15:24:06) [Clang 13.0.0 (clang-1300.0.29.30)]
---Optional dependencies---
pyarrow: 10.0.1
pandas: 1.5.3
numpy: 1.24.1
fsspec: <not installed>
connectorx: <not installed>
xlsx2csv: <not installed>
deltalake: <not installed>
matplotlib: 3.6.3
```

</details>

**Comment by ritchie46** [comment]
Could someone confirm we have a bug?

**Comment by swpease** [comment]
I think the issue is in [`groupby_values_iter_full_lookahead`](https://github.com/pola-rs/polars/blob/64abfa409355e876e3774521d8cdcdfbe8ee9ddf/polars/polars-time/src/windows/groupby.rs#L322), specifically:
```
let slice = unsafe { time.get_unchecked(i..)
let len = slice.partition_point(|v| b.is_member(*v, closed_window));
```
because `slice` does not account for the `offset`, so it is (in general) not partitioned, so `partition_point` doesn't work.

So like, for Example 1, the first `slice` has [1/10, 1/11, 1/12], so `is_member` yields [F, T, F].

**Comment by MarcoGorelli** [comment]
With the latest release, I get
```python
In [6]: from datetime import datetime
   ...: df = pl.DataFrame(
   ...:     {
   ...:         "time": [
   ...:             datetime(2021, 1, 10),
   ...:             datetime(2021, 1, 11),
   ...:             datetime(2021, 1, 12),
   ...:         ],
   ...:         "val": range(1, 4),
   ...:     }
   ...: ).sort('time')
   ...: (df
   ...:     .groupby_rolling("time",
   ...:                      period="1d",
   ...:                      offset="0d",
   ...:                     )
   ...:     .agg([
   ...:         pl.col("val").sum(),
   ...:         pl.col("time").min().alias("min_date"),
   ...:         pl.col("time").max().alias("max_date"),
   ...:     ])
   ...: )
Out[6]:
shape: (3, 4)
┌─────────────────────┬──────┬─────────────────────┬─────────────────────┐
│ time                ┆ val  ┆ min_date            ┆ max_date            │
│ ---                 ┆ ---  ┆ ---                 ┆ ---                 │
│ datetime[μs]        ┆ i64  ┆ datetime[μs]        ┆ datetime[μs]        │
╞═════════════════════╪══════╪═════════════════════╪═════════════════════╡
│ 2021-01-10 00:00:00 ┆ 2    ┆ 2021-01-11 00:00:00 ┆ 2021-01-11 00:00:00 │
│ 2021-01-11 00:00:00 ┆ 3    ┆ 2021-01-12 00:00:00 ┆ 2021-01-12 00:00:00 │
│ 2021-01-12 00:00:00 ┆ null ┆ null                ┆ null                │
└─────────────────────┴──────┴─────────────────────┴─────────────────────┘
```

which looks correct

The windows are:
- (2021-01-10, 2021-01-11]
- (2021-01-11, 2021-01-12]
- (2021-01-12, 2021-01-13]

**Comment by swpease** [comment]
Yeah, it looks like it's working.

I made a [draft pull request](https://github.com/pola-rs/polars/pull/6790) for this in February. I see in it that I had also made a change to `groupby_values_iter_partial_lookbehind`, with the comment: 

> This change also indicated the need to change `groupby_values_iter_partial_lookbehind` to allow the lagging_offset to go past the current index `i`.

I don't really remember what I was talking about. Um... well I may have just finally dusted off the cobwebs? This method is billed as "correct for all lookbehind/lookaheads, but is slower", so I think I misinterpreted it as meaning it should be capable of lookaheads. Maybe I should've just read the method name 😅.
